tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,292 @@
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          DataType_Name(static_cast<DataType>(data_type)),
+          " is not supported by dlpack");
+      break;
+  }
+  return dtype;
+}
+
+DLContext GetDLContext(TFE_TensorHandle* h, TF_Status* status) {
+  DLContext ctx;
+  const char* device_name = h->handle->DeviceName(&status->status);
+  DeviceNameUtils::ParsedName parsed_name;
+  tensorflow::DeviceNameUtils::ParseFullName(device_name, &parsed_name);
+  std::string device_type = parsed_name.type;
+  int device_id = -1;
+  if (parsed_name.has_id) {
+    device_id = parsed_name.id;
+  }  // Question: Is it possible that it doens't have id?
+
+  ctx.device_id = device_id;
+  if (device_type == "CPU") {
+    ctx.device_type = DLDeviceType::kDLCPU;
+  } else if (device_type == "GPU") {
+    ctx.device_type = DLDeviceType::kDLGPU;
+  } else {
+    status->status = tensorflow::errors::InvalidArgument(
+        "Unsupported Device Type for dlpack");
+  }
+
+  return ctx;
+}
+
+DLManagedTensor* TFEHandleToTFDLManagedTensorCtx(TFE_TensorHandle* h,
+                                                 TF_Status* status) {
+  const Tensor* tensor = GetTensorFromHandle(h, status);
+  TF_DataType data_type = static_cast<TF_DataType>(tensor->dtype());
+  auto* tf_dlm_tensor_ctx = new TFDLManagedTensorCtx;
+
+  TensorReference* tensor_ref =
+      new TensorReference(*tensor);  // This will call buf_->Ref()
+  tf_dlm_tensor_ctx->handle = tensor_ref;
+  tf_dlm_tensor_ctx->tensor.manager_ctx = tf_dlm_tensor_ctx;
+  tf_dlm_tensor_ctx->tensor.deleter = &DLManagedTensorDeleter;
+  tf_dlm_tensor_ctx->tensor.dl_tensor.ctx = GetDLContext(h, status);
+  int ndim = tensor->dims();
+  tf_dlm_tensor_ctx->tensor.dl_tensor.ndim = ndim;
+  tf_dlm_tensor_ctx->tensor.dl_tensor.data =
+      TFE_TensorHandleDevicePointer(h, status);
+  tf_dlm_tensor_ctx->tensor.dl_tensor.dtype = GetDLDataType(data_type, status);
+
+  int64_t* shape_arr = new int64_t[ndim];
+  for (int i = 0; i < ndim; i++) {
+    shape_arr[i] = tensor->dim_size(i);
+  }
+
+  tf_dlm_tensor_ctx->tensor.dl_tensor.shape = shape_arr;
+
+  tf_dlm_tensor_ctx->tensor.dl_tensor.strides = nullptr;
+  tf_dlm_tensor_ctx->tensor.dl_tensor.byte_offset =
+      0;  // TF doesn't handle the strides and byte_offsets here
+  return &tf_dlm_tensor_ctx->tensor;
+}
+
+absl::optional<std::string> DeviceNameFromDlContext(const DLContext& ctx,
+                                                    TF_Status* status) {
+  switch (ctx.device_type) {
+    case DLDeviceType::kDLCPU:
+      return "CPU:0";
+    case DLDeviceType::kDLGPU:
+      return absl::StrCat("GPU:", ctx.device_id);
+    default:
+      return absl::nullopt;
+  };
+}
+TF_DataType TfDataTypeFormDlDataType(const DLDataType& dtype,
+                                     TF_Status* status) {
+  TF_DataType tf_dtype;
+  switch (dtype.code) {
+    case DLDataTypeCode::kDLUInt:
+      switch (dtype.bits) {
+        case 1:
```

Comment:
Isn't this size inconsistent between your export and import functions? Your export size of a bool is 8 bits, whereas your import size is 1 bit. So if nothing else, this won't round trip.

(I couldn't decide the right representation in JAX either, so I made import/export of booleans an error there.)

Reply:
I removed the 1 bits situation here(from_dlpack), but keep the option of exporting bool to DLPack(8 bits). The main reason is some projects(such as DGL) needs boolean types, for example using boolean masks.

Reply:
Resolving this assuming the last comment addresses Peter's concern.